Add host API for *_on_stream operations

amd-arozanov opened the pull request:
## Motivation
(AI Assist)

This PR extends host-side asynchronous stream-based API introduced in https://github.com/ROCm/rocSHMEM/pull/333 and https://github.com/ROCm/rocSHMEM/pull/274 by adding support for additional collective, RMA, signaling, and synchronization operations.

The new `*_on_stream` host functions enable users to enqueue rocSHMEM operations on HIP streams, allowing parallel execution across multiple streams and better overlap of communication and computation.

Ticket: https://amd-hub.atlassian.net/browse/AIROCSHMEM-202

## Technical Details

This PR adds the following host-side stream API functions:

**Collective Operations:**
- `__host__ void rocshmem_barrier_all_on_stream(hipStream_t stream)` - extended support to GDA and RO backends (previously only IPC backend was supported)
- `__host__ void rocshmem_broadcastmem_on_stream(rocshmem_team_t team, void *dest, const void *source, size_t nelems, int pe_root, hipStream_t stream)` - **new**

**RMA Operations:**
- `__host__ void rocshmem_getmem_on_stream(void *dest, const void *source, size_t bytes, int pe, hipStream_t stream)` - **new**
- `__host__ void rocshmem_putmem_on_stream(void *dest, const void *source, size_t bytes, int pe, hipStream_t stream)` - **new**

**Signaling Operations:**
- `__host__ void rocshmem_putmem_signal_on_stream(void *dest, const void *source, size_t bytes, uint64_t *sig_addr, uint64_t signal, int sig_op, int pe, hipStream_t stream)` - **new**
  - Added native `uint64_t` support to wait_until API (generated via `P2P_SYNC.py`). Added device functions: `rocshmem_uint64_wait_until`, `rocshmem_uint64_wait_until_any`, `rocshmem_uint64_wait_until_all`, `rocshmem_uint64_wait_until_some`, and their vector variants

**Point-to-Point Synchronization:**
- `__host__ void rocshmem_signal_wait_until_on_stream(uint64_t *sig_addr, int cmp, uint64_t cmp_value, hipStream_t stream)` - **new**

**Implementation Details:**

- All functions are implemented by launching GPU kernels that call existing device-side rocSHMEM workgroup collective functions
- Collective operations (`broadcastmem_on_stream`) create a separate context for each workgroup via `rocshmem_wg_team_create_ctx()` to avoid contention on the default context, allowing parallel execution across multiple streams
- Support added across all backends (GDA, IPC, RO)

**Testing:**

- Added functional tests for all new APIs:
  - `barrier_all_on_stream_tester`
  - `team_broadcastmem_on_stream_tester`
  - `getmem_on_stream_tester`
  - `putmem_on_stream_tester`
  - `putmem_signal_on_stream_tester`
  - `signal_wait_until_on_stream_tester`

## Test Plan

rocSHMEM functional testing

## Test Result

All tests passed except:
- teamreduction_n2_w1_z1_32768B. It fails in other PRs too


## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

amd-arozanov committed:
Add functional test for barrier_all_on_stream

amd-arozanov committed:
Add rocshmem_barrier_all_on_stream support for GDA and RO backends

Implements rocshmem_barrier_all_on_stream operation for
GPU Direct Access and Reverse Offload backends.

Previously, rocshmem_barrier_all_on_stream was only supported for IPC backend.

amd-arozanov committed:
Add functional test for rocshmem_broadcastmem_on_stream

amd-arozanov committed:
Add host-side rocshmem_broadcastmem_on_stream API

Implement stream-based broadcast collective operation

- Add rocshmem_broadcastmem_on_stream host API and kernel implementation
- Add functional test TeamBroadcastmemOnStreamTester with multi-stream
  support and correctness verification
- Use per-workgroup contexts to avoid contention across parallel streams

API:
rocshmem_broadcastmem_on_stream(team, dest, source, nelems, pe_root, stream)

amd-arozanov committed:
Add functional test for rocshmem_getmem_on_stream

amd-arozanov committed:
Add host-side rocshmem_getmem_on_stream API

Implement stream-based point-to-point RMA get operation

- Add rocshmem_getmem_on_stream host API and kernel implementation
- Support for asynchronous getmem operations on HIP streams
- Add backend support for GDA, RO, and IPC contexts
- Use work-group collective getmem for efficient memory transfer

API:
rocshmem_getmem_on_stream(dest, source, nelems, pe, stream)

(AI Assist)

amd-arozanov committed:
Add host-side rocshmem_putmem_on_stream API

- Add rocshmem_putmem_on_stream for asynchronous remote writes
- Support for concurrent RMA operations on HIP streams
- Add backend support for GDA, RO, and IPC contexts
- Use work-group device collective operation

API:
rocshmem_putmem_on_stream(dest, source, bytes, pe, stream)

(AI Assist)

amd-arozanov committed:
Add functional test for rocshmem_putmem_on_stream

amd-arozanov committed:
Add host-side rocshmem_putmem_signal_on_stream API

Enables asynchronous putmem operations with signaling on HIP streams.

The implementation includes:
- Kernel wrapper rocshmem_putmem_signal_kernel
- Host interface putmem_signal_on_stream method
- Context layer support across all backends (IPC, GDA, RO)
- Public API

Function signature:
void rocshmem_putmem_signal_on_stream(void *dest, const void *source,
                                      size_t bytes, uint64_t *sig_addr,
                                      uint64_t signal, int sig_op,
                                      int pe, hipStream_t stream);

amd-arozanov committed:
Add functional test for rocshmem_putmem_signal_on_stream

amd-arozanov committed:
Add host-side rocshmem_signal_wait_until_on_stream API

Enables asynchronous signal wait operations on HIP streams.

The implementation includes:
- Kernel wrapper rocshmem_signal_wait_until_kernel
- Host interface signal_wait_until_on_stream method
- Context layer support across all backends (IPC, GDA, RO)
- Native uint64_t support in wait_until API (generated from P2P_SYNC.py)

Function signature:
void rocshmem_signal_wait_until_on_stream(uint64_t *sig_addr, int cmp,
                                          uint64_t cmp_value,
                                          hipStream_t stream);

(AI Assist)

amd-arozanov committed:
Add functional test for rocshmem_signal_wait_until_on_stream

amd-arozanov committed:
Add documentation for stream API functions

This commit adds API documentation for the following host-side
stream functions:

- rocshmem_barrier_all_on_stream (collective routines)
- rocshmem_broadcastmem_on_stream (collective routines)
- rocshmem_getmem_on_stream (RMA operations)
- rocshmem_putmem_on_stream (RMA operations)
- rocshmem_putmem_signal_on_stream (signaling operations)
- rocshmem_signal_wait_until_on_stream (point-to-point sync)

The documentation includes function signatures, parameter descriptions,
and detailed explanations of asynchronous behavior and stream handling.

(AI Assist)

Yiltan commented on docs/api/rma.rst:
```diff
@@ -67,6 +67,27 @@ ROCSHMEM_PUTMEM
+ROCSHMEM_PUTMEM_ON_STREAM
+--------------------------
+
+.. cpp:function:: __host__ void rocshmem_putmem_on_stream(void *dest, const void *source, size_t bytes, int pe, hipStream_t stream)
```
This API should be keep consistent with `rocshmem_putmem`. There are a few comments in the docs adjusting `bytes` to `nelems`, the same changes should be made to other APIs and the code itself.

I agree with what was done here with `bytes` (it makes more sense to me too) but I think its best to stick with convention



```suggestion
.. cpp:function:: __host__ void rocshmem_putmem_on_stream(void *dest, const void *source, size_t nelems, int pe, hipStream_t stream)
```

amd-arozanov replied:
Done

Yiltan commented on docs/api/rma.rst:
```diff
@@ -67,6 +67,27 @@ ROCSHMEM_PUTMEM
+
+    :param dest:    Destination address. Must be an address on the symmetric heap.
+    :param source:  Source address. Must be an address on the symmetric heap.
+    :param bytes:   Number of bytes to transfer.
```
```suggestion
    :param nelems:   The number of elements to transfer.
```

amd-arozanov replied:
Done. I took the wording from `rocshmem_putmem`. It clarifies that `nelems` is in bytes

":param nelems: Size of the transfer in bytes."

Yiltan commented on docs/api/rma.rst:
```diff
@@ -67,6 +67,27 @@ ROCSHMEM_PUTMEM
+
+**Description:**
+This routine enqueues a putmem RMA operation on a HIP stream. The function writes contiguous
+data of ``bytes`` bytes from source on the calling PE to ``dest`` at ``pe``. The operation
```
```suggestion
data of ``nelems`` elements from source on the calling PE to ``dest`` at ``pe``. The operation
```

amd-arozanov replied:
In `ROCSHMEM_PUTMEM` description it states "``nelems`` bytes". I would prefer also to keep the same for `ROCSHMEM_PUTMEM_ON_STREAM` and others, because it clarifies for the customer that one element is one byte

Yiltan commented on tests/functional_tests/putmem_signal_on_stream_tester.cpp:
```diff
@@ -0,0 +1,235 @@
+  n_pes = rocshmem_n_pes();
+
+  char *value{nullptr};
+  if ((value = getenv("ROCSHMEM_NUM_STREAMS"))) {
```
```suggestion
  if ((value = getenv("ROCSHMEM_TEST_NUM_STREAMS"))) {
```

Add the `_TEST_` infix to signify this is not for the rocSHMEM library but rather the tests. Same of other environment variables in this PR. We have a similar environment variable `ROCSHMEM_TEST_UUID`

amd-arozanov replied:
Done

Yiltan reviewed (commented):
Looks good, very minor comments.

amd-arozanov committed:
Rename "bytes" -> "nelems"

amd-arozanov committed:
Add "_TEST_" to the variables used in tests

edgargabriel commented on src/host/host.cpp:
```diff
@@ -364,6 +365,133 @@ __host__ void HostInterface::alltoallmem_on_stream(rocshmem_team_t team,
+  // launch kernel to do broadcast with given stream, if none, use default
+  // stream
+  if (stream == nullptr) {
+    stream = hipStreamDefault;
```
I don't think this is correct. This is a common misconception that `hipStreamDefault` is the default stream. Its not. Its a value for `flags` in hipStreamCreateWithFlags. The default Stream has a value of `0` as far as I know. See 

https://rocm.docs.amd.com/projects/HIP/en/latest/doxygen/html/group___stream.html#gaf2382e3cc6632332a8983a0f58e43494

amd-arozanov replied:
Thank you for finding this! I removed hipStreamDefault. Yes, the name hipStreamDefault is confusing.

stream == nullptr should work as the default stream.

edgargabriel reviewed (commented):
Other than the one comment, it looks good, thank you!

abouteiller commented on docs/api/pt2pt_sync.rst:
```diff
@@ -96,6 +96,27 @@ ROCSHMEM_TEST
+ROCSHMEM_SIGNAL_WAIT_UNTIL_ON_STREAM
+-------------------------------------
+
+.. cpp:function:: __host__ void rocshmem_signal_wait_until_on_stream(uint64_t *sig_addr, int cmp, uint64_t cmp_value, hipStream_t stream)
```
```suggestion
.. cpp:function:: __host__ void rocshmem_signal_wait_until_on_stream(uint64_t *ivars, int cmp, uint64_t val, hipStream_t stream)
```
(there are a number of these): the names should be consistent with the rest of the documentation for the non-stream procedures. To be fair `cmp_value` is a better name and maybe amending the pre-existing procedures to match would be a good alternative. `sig_addr` is for signal_op procedures only not for wait_until and friends.

amd-arozanov replied:
Thank you for finding this! I renamed the arguments to be consistent with other functions.

abouteiller replied:
This applies only to wait_until not signal_wait_until, so the original form was correct.

abouteiller commented on include/rocshmem/rocshmem.hpp:
```diff
@@ -365,6 +365,102 @@ __host__ void rocshmem_alltoallmem_on_stream(rocshmem_team_t team, void *dest,
+ *
+ * @return void
+ */
+__host__ void rocshmem_signal_wait_until_on_stream(uint64_t *sig_addr, int cmp,
```
Same comment about sig_addr and cmp_value

abouteiller replied:
Other that these minor things it looks good thanks!

amd-arozanov replied:
Done

amd-arozanov committed:
Remove incorrect hipStreamDefault usage

hipStreamDefault is not a default stream. This is a flag.

If stream == nullptr, then just pass it to kernel. It will launch the kernel on the default stream
